### DISCUSSION TITLE
Stability issues with my google server

### DISCUSSION THREAD

**Discussion by Ciancior** [mikejac/node-red-contrib-google-smarthome#602] [Q&A] [answered]
Good morning everyone, I have been using this Add for about two years, just configured on a Rasberry pi4, everything went well, despite my 100 devices on/off, everything was reliable and disconnections from the Google server were really rare. I should point out that I have an ADSL connection, with a dynamic IP, but I get a certified domain thanks to the use of duckdns.org and a reverse_proxy.
I had line problems for about six months, and when I tried to turn on one of the lights, Google often told me that it could not reach the server in my house.
Solved the problems with the line, about a month ago, doing the various external checks of the certificate and connecting with https://example.com:3001/check, and everything is ok, but unfortunately the problem with the functioning remained, when I asked to turn on a light, Google replied that the server in my house is not reachable, and often on the Google app all the devices connected to it are offline.
In addition, I have never been able to use local fulfillment, which could have solved the problem, but despite a thousand attempts I have never managed to activate it. Please, can someone help me?

**Comment by Caprico85** [comment, marked as answer]
Hi @Ciancior,

Could it be DNS issues?

I had similar problems. For me, they were related to DNS. I checked by repeatedly doing `nslookup [mydomain.org](http://mydomain.org/)` and with online services like [nslookup.io](http://nslookup.io/). Multiple times in a row, multiple times a day.

Sometimes I got the correct IP address for my domain. Sometimes my domain was resolved to an old IP address I had before the last reconnect. Sometimes I got no results at all. Sometimes I got only the IPv6 or IPv4 address, but not the other. Often enough, checking 10 times in a row got me 3 different results. And sometimes, updating my IP address failed too.

I switched DNS providers several times. I tried with freedns.afraid.org, Dynv6 and DuckDNS. All of them had various problems. Since then, I don't really trust free DNS providers anymore.

After months of experimenting I'm currently on the following setup. Worked stable for the last 4-5 months.

- A domain from [eu.org](http://eu.org/). Their domains allow you to set the NS record for your domain. This allows you to use any nameserver you want. Unfortunately, [eu.org](http://eu.org/) is not registering new domains anymore.
- As the nameserver I'm using Cloudflare. They have a free tier and they allow you to set the DNS records for your domains using their API. I'm not using their proxying and tunneling, I only use the nameserver. And I trust Cloudflare a little bit more than all the other free dyndns provider
- As the DNS updater I'm using https://github.com/qdm12/ddns-updater.

**Reply by Ciancior** [reply]
Hi @Caprico85,
I suspected that was the problem!!!! I will try to work in this direction.

Were you able to configure local fulfillment? Do you know if it is something that works on this Add? I tried in every way with all the configuration variants, but nothing worked!!!

Do you have any advice for me?

Thanks in advance.

**Reply by Caprico85** [reply]
Yes, local fulfillment works for me.

Local execution is rather network dependent. Can you tell me a bit more about your setup first? Where do you run Node-RED? In a Docker container? On a Raspberry Pi, a NAS or something? Any specialties in your network like separated IOT/guest networks, firewalls or something?

Can you also show screenshopts of local fulfillment config in Node-RED

![image](https://github.com/user-attachments/assets/90a62d75-7967-4b6f-87ba-17c8c831e041)

and in the Actions Console

![image](https://github.com/user-attachments/assets/5dd55b4d-aebc-4290-b000-072cdb90585e)

What do you see in `chrome://inspect`?

![image](https://github.com/user-attachments/assets/25bad86b-d0bc-4236-b55a-54b4cacf9871)

**Reply by Ciancior** [reply]
Hi Caprico85, sorry if I'm only answering you now, but I had covid again....
The Node-Red is installed on a Raspberry pi4, with the Adds: Node-red-contributi-google-smarthome, Node-red-dashboard, and Node-red-contrib-s7.
I have a Dashboard that I view locally from smartphones and tablets, to control and view all the virtual devices shared with Node-red-contributi-google-smarthome, which represent the tags of the six Siemens S7 1200, which is the entire home automation system of my house, (over 120 digital outputs, and 200 digital inputs).
I have a Dynamic IP, but I get a certified domain thanks to the use of duckdns.org and a reverse_proxy.
In the local network I have a Multi Van Tp-link er7206, which controls both ADSL and the G4 modem, but which is only a backup.
The Tp-link er7206, clearly has the open ports, incoming and outgoing 80, 443 and 3001 to the address of the Raspberry (192.168.1.9) on which the Node-red is running and the Caddy that takes care of renewing the SSL certificate.
I have a mesh wifi network, with 4 devices that spread the network in the house and on which the 14 google nest hubs and the 10 google chromecasts are connected.
The Raspberry is connected to the lan 192.168.1.9, and to the wifi 192.168.3.9, and both ip are static.
This evening I will be able to see What do I see in chrome://inspect?, and I will send you any errors.

Thank you very much for your interest.

[Properties_Node-red.pdf](https://github.com/user-attachments/files/17633141/Properties_Node-red.pdf)
[Actions_Console.pdf](https://github.com/user-attachments/files/17633142/Actions_Console.pdf)

**Reply by Caprico85** [reply]
Oh wow, this s an incredible setup. I have 1 RasPi, 3 lamps, 1 thermostat and that's it.

One thing that I noticed in your screenshots is the broadcast address set in the Actions console.

![image](https://github.com/user-attachments/assets/53455708-8866-4225-bc9a-94644b575f7e)

With local fulfilment, your smart speakers regularly send out UDP packets to the network specified by the broadcast address, trying to get a response from the RasPi. Your smart speakers are on the wifi network 192.168.3.255. As they are currently configured, they will try to reach the RasPi on the lan network 192.168.1.255. This can be problematic as UDP packets are not usually routed between different networks. Since your RasPi is also connected to the wifi network, the best way for them to communicate is the wifi network 192.168.3.255.

Long story short, set your broadcast address to 192.168.3.255. This may already fix local fulfilment.

**Reply by Ciancior** [reply]
Thanks Caprico85,
I'll try to do it right away, as soon as I'm home, and in any case I'll query chrome://inspect, to see if there are any logs with errors.
If I could solve the problem with the stability of the certified domain, but above all the problem of local fulfillment, I would like to develop this system even more, because I have 6 Siemens PLCs that manage the whole house in a robust and hyper-reliable way, with a program written by me, which I'm 100% sure of, and then with the Node-red I interface them to a Dashboard that I manage as I want, multimedia, weather forecast, garden, alarms, heating..... and to Google Home, thanks to your wonderful Add.

P.S. How long does local fulfillment take on average to be active?
Which protocol is more suitable in your opinion, UDP or MDNS?

**Reply by Ciancior** [reply]
I applied, the new broadcast address to 192.168.3.255, but it still can't go locally....
I'll send you a pdf of the chrome://inspect.....
Please tell me if I can do something else.....

Thanks again

[Inspect.pdf](https://github.com/user-attachments/files/17637272/Inspect.pdf)
[Ispect2.pdf](https://github.com/user-attachments/files/17642519/Ispect2.pdf)
[Ispect3.pdf](https://github.com/user-attachments/files/17642520/Ispect3.pdf)

**Reply by Caprico85** [reply]
It's a good sign that you are getting the REACHABLE_DEVICES messages. This is the smart speaker communicating with Node-RED on which devices it can reach. That means, the local connection has been successfully established.

Local fulfilment should be already working by now. Unfortunately, it's now up to Google to decide wether or not to use the local connection. Sometimes it decides that it doesn't want to use local fulfillment, even if it's set up correctly. Google usually doesn't give any details on this.

A few things I would try now:
- Assign your devices to rooms in the Google Home app. I've read somewhere that unassigned devices may not use local fulfillment.
- Say "Hey Google, force local" (english command works on non-english devices too). This forces Google to only use the local connection and fail otherwise. If local fulfillment is the only chance to execute commands, maybe this gets Google to actually use local fulfillment. Revert back to default (local with fallback to cloud) by saying "Hey Google, force default".
- If you're looking for the ring icon on the devices nodes, disconnect any inputs from your device nodes. It's possible that another node is sending inputs to the Google device node, immediately overriding the ring icon.- - Restart your smart speaker(s) and Node-RED.
- Wait. For days. Or even weeks. I've seen it happen more than once that things started working days later out of the blue after desperately trying to get it to work and already giving up.

**Reply by Ciancior** [reply]
Thanks for all your suggestions, I will wait patiently.
In the meantime the possibility of connecting the optical fiber directly to my house has arrived, and to have a fixed IP.
I bought a domain, and I hope to make everything more stable.
